pcsx2/Recording/InputRecordingFile.cpp

```diff
@@ -0,0 +1,464 @@
+	if (savestate.fromSavestate) {
+		// We read the size (and the savestate) only if we must
+		if (fread(&savestate.savestatesize, sizeof(savestate.savestatesize), 1, fp) != 1) return false;
+		if (savestate.savestatesize == 0) {
+			recordingConLog(L"[REC]: Invalid size of the savestate.\n");
+			return false;
+		}
+
+		savestate.savestate.MakeRoomFor(savestate.savestatesize);
+		// We read "savestatesize" * the size of a cell
+		if (fread(savestate.savestate.GetPtr(), sizeof(savestate.savestate[0]), savestate.savestatesize, fp)
+			!= savestate.savestatesize) return false;
+
+		// We load the savestate
+		memLoadingState load(savestate.savestate);
+		UI_DisableSysActions();
+		GetCoreThread().Pause();
+		SysClearExecutionCache();
+		load.FreezeAll();
+		GetCoreThread().Resume();
+	}
+	else {
+		sApp.SysExecute();
+	}
+
+	// ID
+	if (header.ID != 0xCC) {
+		return false;
+	}
+	// ver
+	if (header.version != 3) {
+		return false;
+	}
+	return true;
+}
+bool InputRecordingFile::writeHeader()
+{
+	if (fp == NULL)return false;
+	rewind(fp);
+	if (fwrite(&header, sizeof(InputRecordingHeader), 1, fp) != 1) return false;
+	return true;
+}
+bool InputRecordingFile::writeSavestate()
+{
+	if (fp == NULL) return false;
+	fseek(fp, SEEKPOINT_SAVESTATE, SEEK_SET);
+	if (fwrite(&savestate.fromSavestate, sizeof(bool), 1, fp) != 1) return false;
+
+	if (savestate.fromSavestate) {
+		if (fwrite(&savestate.savestatesize, sizeof(savestate.savestatesize), 1, fp) != 1) return false;
+		if (fwrite(savestate.savestate.GetPtr(), sizeof(savestate.savestate[0]), savestate.savestatesize, fp)
+			!= savestate.savestatesize) return false;
+	}
+	return true;
+}
+bool InputRecordingFile::writeMaxFrame()
+{
+	if (fp == NULL)return false;
+	fseek(fp, SEEKPOINT_FRAMEMAX, SEEK_SET);
+	if (fwrite(&MaxFrame, 4, 1, fp) != 1) return false;
+	return true;
+}
+void InputRecordingFile::updateFrameMax(unsigned long frame)
+{
+	if (MaxFrame >= frame) {
+		return;
+	}
+	MaxFrame = frame;
+	if (fp == NULL)return ;
+	fseek(fp, SEEKPOINT_FRAMEMAX, SEEK_SET);
+	fwrite(&MaxFrame, 4, 1, fp);
+}
+void InputRecordingFile::addUndoCount()
+{
+	UndoCount++;
+	if (fp == NULL)return;
+	fseek(fp, SEEKPOINT_UNDOCOUNT, SEEK_SET);
+	fwrite(&UndoCount, 4, 1, fp);
+
+}
+
+void InputRecordingHeader::setAuthor(wxString _author)
+{
+	int max = ArraySize(author) - 1;
+	strncpy(author, _author.c_str(), max);
+	author[max] = 0;
+}
+void InputRecordingHeader::setCdrom(wxString _cdrom)
+{
+	int max = ArraySize(cdrom) - 1;
+	strncpy(cdrom, _cdrom.c_str(), max);
+	cdrom[max] = 0;
+}
+void InputRecordingHeader::init()
+{
+	memset(author, 0, ArraySize(author));
+	memset(cdrom, 0, ArraySize(cdrom));
+}
+
+void InputRecordingFile::ConvertV2ToV3(wxString filename)
+{
+	// TODO, with the latest version of save states requiring a savestate integrated into the file
+	// or it restarts as it assumes it is from power-on
+	// I don't see how save states would be compatible
```

Comment:
These options were added in the hopes of supporting breaking changes, however I've now removed them as I think this is a good opportunity to start from "fresh" (and half of them were not even implemented / working correctly).  The conversions were flawed since savestates are now stored along with the movie file, it made supporting conversions cross pcsx2 versions difficult (legacy = pcsx2 from 9 years ago).  I think that there is still value in supporting these conversions but I will break the work out into a separate tool / script.

In regards to whether it is useful to provide it within PCSX2...I think I would argue yes.  Simply because it is probably the best way to distribute it to users, but I suppose that is a concern that can be ironed out later in the event that another breaking change has to be made.  Maybe it all living in an external tool would be fine, and perhaps just have a link from the GUI to it?